Refactor get analytical jacobian

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54480 Implement faster gradcheck and enable for all ops except some
* #54479 Factor out numerical logic
* #54378 Refactor get numerical jacobian to calculate wrt all outputs at once
* #54092 Refactor get numerical jacobian
* **#54049 Refactor get analytical jacobian**

### For release notes:
`torch.autograd.gradcheck.get_analytical_jacobian` (not part of the public api) is being deprecated.

--- end --

The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)

Differential Revision: [D27307240](https://our.internmc.facebook.com/intern/diff/D27307240)

soulitzer committed:
Refactor get analytical jacobian

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -161,13 +161,54 @@ def get_stride(size):
+        return torch.autograd.grad(output, diff_input_list, grad_output,
+                                   retain_graph=True, allow_unused=True)
+    fn = custom_backward_fn if custom_backward_fn is not None else backward_fn
+    jacobians_rows = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
+    jacobians_rows_reentrant = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
```
I never really understood what the reentrant check is testing. What is it testing?

The reason why I am asking is because previously, we would compute row 0 of the first jacobian, then row 0 of the second jacobian, then row 1 of the first jacobian, then row 1 of the second jacobian, etc. This PR changes it so that we compute the full first jacobian followed by the full second jacobian. Does this affect the reentrant check?

soulitzer replied:
I don't really know what its named 'reentrant' check, AFAICT its just computing something twice and checking if they are the same as to see if its deterministic. A better name imo should be determinism check not reentrant check. Maybe @albanD knows more.

zou3519 replied:
Offline @albanD said this is re-entrant in the sense that it is calling the same function twice. This is different from when we call autograd from within autograd, which we also refer to as "re-entrant". So yes, looks like a case of bad naming

zou3519 replied:
Also I realized that the code does explain this in an error message: https://github.com/pytorch/pytorch/blob/87989a6cf9979d3842991ab1696a62a1d247a523/torch/autograd/gradcheck.py#L472-L475

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
+
+def get_analytical_jacobian(fn, sample_output, grad_out_scale):
+    # Computes Jacobian row-by-row using backward function `fn` = v^T J
+    # NB: we can't combine the rows into a single jacobian tensor because fn(v) for
+    # different v may return tensors with different number of elements
```
Is this because we expect fn(v) to return tensors with the same number of elements if the gradient formula is correct but are not making that assumption because it can be wrong?

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
+    return jacobians, failed
+
+
+def get_analytical_jacobian(fn, sample_output, grad_out_scale):
```
nit: fn -> vjp_fn is more descriptive (v^T J is a vector-jacobian-product)

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
+    grad_out_base = torch.zeros_like(sample_output, memory_format=torch.legacy_contiguous_format)
+    flat_grad_out = grad_out_base.view(-1)
+    # jacobians_rows[i][j] represents the jth row of the ith input
+    jacobians_rows: Dict[int, List[Optional[torch.Tensor]]] = {}
```
Why does it need to be a Dict[int, List[Optional[torch.Tensor]]]? Couldn't it be a List[List[Optional[torch.Tensor]]]?

soulitzer replied:
Yeah, its probably safe to assume that different grad_outs to backward produce the same number of inputs.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -161,13 +161,54 @@ def get_stride(size):
+            else:
+                row_dense = row.to_dense() if not row.layout == torch.strided else row
+                assert out_jacobian[:, j].numel() == row_dense.numel()
+                out_jacobian[:, j] = row_dense.contiguous().view(-1)
```
nit: row_dense.reshape(-1) does the same thing

zou3519 reviewed (commented):
refactor looks correct, I had a few questions

soulitzer committed:
Update on "Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)



[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get analytical jacobian"


The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)

Differential Revision: [D27307240](https://our.internmc.facebook.com/intern/diff/D27307240)

[ghstack-poisoned]
